### ISSUE TITLE
Particle marginal Metropolis-Hastings (PMMH) 

### ISSUE THREAD

**Issue by emilemathieu** [TuringLang/Turing.jl#332]
Hi all !

Have you already thought of implementing [PMMH](https://darrenjw.wordpress.com/2011/05/17/the-particle-marginal-metropolis-hastings-pmmh-particle-mcmc-algorithm/) ?
I'm interested in implementing it ! 

The syntax could be the following (with the same notations as in the article):
```julia
sampler1 = PMMH(nb_samples, SMC(1, :x) , :θ)
sampler2 = PMMH(nb_samples, SMC(1, :x) , (:θ, proposal))
```
with `proposal` being a `distribution` type (with `rand` and `pdf` functions available).
Here, `sampler1` would be by default defining `q(θ'|θ)` as the prior and `sampler1` defining `q = proposal `.

**Comment by yebai** [comment]
>Have you already thought of implementing PMMH ?
> I'm interested in implementing it !

Sounds great - we thought about it but didn't have the time to implement it. We'll do our best to help.

>The syntax could be the following (with the same notations as in the article):
>sampler1 = PMMH(nb_samples, SMC(1, :x) , :θ)
>sampler2 = PMMH(nb_samples, SMC(1, :x) , (:θ, proposal))

This looks very reasonable!

**Comment by emilemathieu** [comment]
Hi @yebai !

I'm working on PMMH, and I'm wondering why cannot we use SMC inside a Gibbs sampler (but only PG and Hamiltonian) ? Is it because you didn't adapt SMC (with ` vi[getretain(vi, 0, spl)] = NULL` and such) or is there a theoretical issue ?

**Comment by yebai** [comment]
> I'm working on PMMH, and I'm wondering why cannot we use SMC inside a Gibbs sampler (but only PG and Hamiltonian) ? Is it because you didn't adapt SMC (with vi[getretain(vi, 0, spl)] = NULL and such) or is there a theoretical issue ?


It's because of theoretical reasons. SMC is not a valid MCMC kernel, combining SMC with HMC would lead to a biased MCMC sampler.

**Comment by xukai92** [comment]
@yebai Is this the same reason why PG use Cond-SMC as a MCMC kernel?

**Comment by yebai** [comment]
> @yebai Is this the same reason why PG use Cond-SMC as a MCMC kernel?

yes that's right.

**Comment by emilemathieu** [comment]
Thanks for the reminding @yebai.

Otherwise, could you explain please why `if spl.alg.gid != 0 link!(vi, spl) end ` and ` if spl.alg.gid != 0 invlink!(vi, spl) end` is necessary in HMC ?

**Comment by xukai92** [comment]
@emilemathieu The transformation is performed only when HMC is used in Gibbs (indicated by the condition `spl.alg.gid != 0`). When HMC is used together with other samplers, it is necessary to make sure HMC works in the transformed space (say R). For example, when HMC is used together with PG, PG only works in the original space (say X) but HMC needs to work in the transformed space (R), thus the transformation X -> R and R -> X is needed. 

For standalone HMC, we can just run HMC in transformed space R all the time (there is one transformation done before the iterations if it is this case).

**Comment by emilemathieu** [comment]
Thanks @xukai92 !
Would you have some references about theoretically why is this space transformation is needed ?
Do you whether I should do this in PMMH ?

**Comment by emilemathieu** [comment]
Moreover I have issues running tests, I get: `LoadError: UndefVarError: check_numerical not defined`.
Shouldn't we add in `src/Turing.jl`
```julia
export check_numerical
```
 ?

**Comment by emilemathieu** [comment]
PR for review: https://github.com/yebai/Turing.jl/pull/339

**Comment by emilemathieu** [comment]
I've pushed on https://github.com/yebai/Turing.jl/pull/339 the version allowing for parameters' proposal.

**Comment by emilemathieu** [comment]
I have pushed a version of `PMMH` that handles user-defined proposals in PR https://github.com/yebai/Turing.jl/pull/339.

**Comment by emilemathieu** [comment]
PR https://github.com/yebai/Turing.jl/pull/339 has been merged.

**Comment by emilemathieu** [comment]
Hi all !

In the current implementation of PMMH, a small modification would allow to use it as a PIMH algorithm.
For reminder, PMMH targets p(theta, x_{1:T} | y_{1:T}) whereas PIMH targets p(x_{1:T} | y_{1:T}) (global parameters theta are considered as deterministic and fixed).

It would be contracted as:
```julia
pmmh = PMMH(100, SMC(20, :v1), :v2)
pimh = PMMH(100, SMC(20))
```

The only change in the code is:
```julia
if !isempty(spl.alg.space) vi = model(vi=vi, sampler=spl) end
```
which means that if there is no global parameters, the theta* ~ p(.|theta) proposal is not ran.

Do you think a new `PIMH` type should created for transparency: ?
```julia
pimh = PIMH(100, SMC(20))
```
or is it ok to use `PMMH` for both cases ?

Best,
Emile

**Comment by emilemathieu** [comment]
Eventually, several modifications were needed, I pushed a PR https://github.com/yebai/Turing.jl/pull/354.
Feedbacks are welcomed.

Best,
Emile

**Comment by yebai** [comment]
@emilemathieu Yes for transparency, it would be better to have a typealias `PIMH` referencing `PMMH`.

**Comment by emilemathieu** [comment]
@yebai 
such as `typealias PIMH PMMH` ?
Or something like
```julia
abstract PMH
immutable PIMH <: PMH ...
immutable PMMH <: PMH ...
```
and then construct `PIMH` as a special case of `PMMH` ?

**Comment by emilemathieu** [comment]
PR https://github.com/yebai/Turing.jl/pull/369

**Comment by yebai** [comment]
Closed by #369
